s2and/data.py

```diff
@@ -1609,15 +1616,63 @@ def pair_sampling_to_store(
                 and not self.pair_sampling_balanced_classes
                 and not self.pair_sampling_balanced_homonym_synonym
             ):
-                # Take samples from each block weighted by their len of total samples
-                # sample_size = min(len(possible), sample_size)
-                # blockwise_pairs: Dict[str, List[Tuple[str, str, Union[int, float]]]] = {}
-                # for k in blockwise_sig_pairs.keys():
-                #     block_sample_size = int(sample_size * len(blockwise_sig_pairs[k])/len(possible))
-                #     samples = random_sampling(blockwise_sig_pairs[k], block_sample_size, self.random_seed)
-                #     if(len(samples)>0):
-                #         blockwise_pairs[k] = samples
-                return blockwise_sig_pairs, blockwise_cluster_ids
+                # Subsampling, while maintaining transitivity
+                sample_size = min(len(possible), sample_size)
+                pairs_ids = random_sampling(possible, sample_size, self.random_seed)
+                # Construct hashset to make queries faster
+                subsample_id_set = set()
+                for tuple in pairs_ids:
+                    id1, id2, _ = tuple
+                    subsample_id_set.add(id1)
+                    subsample_id_set.add(id2)
+
+                # if not os.path.exists("our_data_subsample.pkl"):
+                #     with open('our_data_subsample.pkl', 'wb') as f:
+                #         pickle.dump(subsample_id_set, f)
+
+                # Remove sig_pairs which are not subsampled
+                for block_id, signatures in blocks.items():
+                    sig_pairs = blockwise_sig_pairs[block_id]
+                    cluster_ids = blockwise_cluster_ids[block_id]
+                    n = len(signatures)
+                    block_len = len(sig_pairs)
+
+                    all_idxs = np.arange(0, n)
+                    sig_idxs_to_keep = []
+                    for i, s in enumerate(signatures):
+                        if (s in subsample_id_set):
+                            sig_idxs_to_keep.append(i)
+
+                    sig_idxs_to_keep = np.sort(sig_idxs_to_keep)
+                    if(sig_idxs_to_keep.size == 0):
```

Comment:
If `sig_idxs_to_keep == 0`, then you should just delete that block from your blockwise_* dicts and add a `continue` statement.

Reply:
Additionally, you can add a check to see if `len(sig_idxs_to_keep) == n` and simply call `continue`, because that block should remain unchanged, right?

Reply:
True, updating that